### DISCUSSION TITLE
Is laravel-data Laravel dependent?

### DISCUSSION THREAD

**Discussion by rendesadam** [spatie/laravel-data#633] [Q&A] [answered]
Is it possible to use this package in a non-Laravel project, for example in Slim framework? 
I like the functionality, but I'm not sure I can use it.

(Of course, I see the name of the package, but I don't see any huge dependency in composer.json)

**Comment by czuniga9** [comment, marked as answer]
It's possible but not very easy. It doesn't have many dependencies but it implicitly depends on a global `app` function which returns a laravel container (package `illuminate/container`) and a global `config` function that returns a config object (package `illuminate/config`) or classes that have similar functions (since it doesn't check for an interface).

I was able to make a small proof of concept only using laravel-data and the container package:

The global helpers (copied from a laravel project):
```php
<?php
use Illuminate\Container\Container;

function app($abstract = null, array $parameters = [])
{
    if (is_null($abstract)) {
        return Container::getInstance();
    }
    return Container::getInstance()->make($abstract, $parameters);
}

function config($key = null, $default = null)
{
    if (is_null($key)) {
        return app('config');
    }
    if (is_array($key)) {
        return app('config')->set($key);
    }
    return app('config')->get($key, $default);
}

```

For the config I made a small class but you'll need the full package for configuring laravel-data since it deals with nested arrays:
```php
class Config
{
    public function __construct(public array $config = []) {}

    public function get(string $key, mixed $default = null): mixed
    {
        return $this->config[$key] ?? $default;
    }

    public function set(array|string $key, mixed $value = null)
    {
        if (is_array($key)) {
            foreach ($key as $_key => $_value) {
                $this->config[$_key] = $_value;
            }
        } else {
            $this->config[$key] = $value;
        }
    }
}
```

The Test data class:

```php
use Spatie\LaravelData\Data;

class Foo extends Data
{
    public function __construct(
        public string $title = 't',
        public string $slug = 's',
    ) {}
}
```

The container configuration:
```php
use Spatie\LaravelData\Support\DataConfig;

app()->singleton('config', function () {
    return new Config([
        'data.normalizers' => [
            \Spatie\LaravelData\Normalizers\ModelNormalizer::class,
            \Spatie\LaravelData\Normalizers\ArrayableNormalizer::class,
            \Spatie\LaravelData\Normalizers\ObjectNormalizer::class,
            \Spatie\LaravelData\Normalizers\ArrayNormalizer::class,
            \Spatie\LaravelData\Normalizers\JsonNormalizer::class,
        ]
    ]);
});
app()->singleton(DataConfig::class, fn () => new DataConfig([]));
```

And finally the test itself:
```php
<?php
$foo = Foo::from(['title' => '123', 'slug' => '456']);
var_dump($foo->title); // "123"
var_dump($foo->slug); // "456"
```

I didn't try anything more complex so you'll probably need to tinker with it to get it to work well.

**Reply by rendesadam** [reply]
Thanks for your comment, very useful!

**Reply by rubenvanassche** [reply]
Yeah something for v6 maybe haha, it will be a lot of work to do it correct but at a certain point in time we'll maybe release spatie/data 😄

**Reply by rubenvanassche** [reply]
Good comment BTW 👍
